src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java

```diff
@@ -185,4 +192,43 @@ public ResponseEntity<Map<Long, Optional<ProgrammingSubmission>>> getLatestPendi
         }));
         return ResponseEntity.ok(pendingSubmissions);
     }
+
+    /**
+     * Resets the specified repository to either the exercise template or graded participation
+     *
+     * @param participationId the id of the programming participation that should be resetted
+     * @param gradedParticipationId optional parameter that specifies that the repository should be set to the graded participation instead of the exercise template
+     * @return the ResponseEntity with status 200 (OK)
+     */
+    @PutMapping("/programming-exercise-participations/{participationId}/reset-repository")
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<Void> resetRepository(@PathVariable Long participationId, @RequestParam(required = false) Long gradedParticipationId)
+            throws GitAPIException, IOException {
+        ProgrammingExerciseStudentParticipation participation = programmingExerciseStudentParticipationRepository.findByIdElseThrow(participationId);
+        ProgrammingExercise exercise = programmingExerciseRepository.findByStudentParticipationIdWithTemplateParticipation(participationId)
+                .orElseThrow(() -> new EntityNotFoundException("Programming Exercise for Participation", participationId));
+        participation.setProgrammingExercise(exercise);
+        if (!programmingExerciseParticipationService.canAccessParticipation(participation) || participation.isLocked()) {
+            throw new AccessForbiddenException("participation", participationId);
+        }
+        if (exercise.isExamExercise()) {
+            throw new BadRequestAlertException("Cannot reset repository in an exam", ENTITY_NAME, "noRepoResetInExam");
+        }
+
+        VcsRepositoryUrl sourceURL;
+        if (gradedParticipationId != null) {
+            ProgrammingExerciseStudentParticipation gradedParticipation = programmingExerciseStudentParticipationRepository.findByIdElseThrow(gradedParticipationId);
+            if (!programmingExerciseParticipationService.canAccessParticipation(participation)) {
```

Comment:
isn't this check redundant because it is the caught in the check above? Line 211

Reply:
Did you mean?
```suggestion
            if (!programmingExerciseParticipationService.canAccessParticipation(gradedParticipation)) {
```